common/flatpak-dir.c

```diff
@@ -3306,130 +3320,186 @@ add_appdata_to_deploy_data (GVariantBuilder *metadata_builder,
-                           g_variant_new_variant (g_variant_new_strv (previous_ids, -1)));
+    g_variant_dict_insert_value (&metadata_dict, "previous-ids",
+                                 g_variant_new_strv (previous_ids, -1));
 
-  add_appdata_to_deploy_data (&metadata_builder, deploy_dir, id);
+  add_commit_metadata_to_deploy_data (&metadata_dict, commit_metadata);
+  add_metadata_to_deploy_data (&metadata_dict, metadata);
+  add_appdata_to_deploy_data (&metadata_dict, deploy_dir, id);
 
   res = g_variant_ref_sink (g_variant_new ("(ss^ast@a{sv})",
                                            origin,
                                            commit,
                                            subpaths ? subpaths : empty_subpaths,
                                            GUINT64_TO_BE (installed_size),
-                                           g_variant_builder_end (&metadata_builder)));
+                                           g_variant_dict_end (&metadata_dict)));
   return g_variant_get_data_as_bytes (res);
 }
 
 static GBytes *
-upgrade_deploy_data (GBytes *deploy_data,
-                     GFile *deploy_dir,
-                     FlatpakDecomposed *ref)
+upgrade_deploy_data (GBytes             *deploy_data,
+                     GFile              *deploy_dir,
+                     FlatpakDecomposed  *ref,
+                     OstreeRepo         *repo,
+                     GCancellable       *cancellable,
+                     GError            **error)
 {
   VarDeployDataRef deploy_ref = var_deploy_data_from_bytes (deploy_data);
   g_autoptr(GVariant) metadata = g_variant_ref_sink (var_metadata_peek_as_gvariant (var_deploy_data_get_metadata (deploy_ref)));
-  GVariantBuilder metadata_builder;
+  g_auto(GVariantDict) metadata_dict = FLATPAK_VARIANT_DICT_INITIALIZER;
   g_autofree const char **subpaths = NULL;
   g_autoptr(GVariant) res = NULL;
   int i, n, old_version;
 
-  g_variant_builder_init (&metadata_builder, G_VARIANT_TYPE ("a{sv}"));
-
-  g_variant_builder_add (&metadata_builder, "{s@v}", "deploy-version",
-                         g_variant_new_variant (g_variant_new_int32 (FLATPAK_DEPLOY_VERSION_CURRENT)));
+  g_variant_dict_init (&metadata_dict, NULL);
+  g_variant_dict_insert_value (&metadata_dict, "deploy-version",
+                               g_variant_new_int32 (FLATPAK_DEPLOY_VERSION_CURRENT));
 
   /* Copy all metadata except version from old */
   n = g_variant_n_children (metadata);
   for (i = 0; i < n; i++)
     {
-      g_autoptr(GVariant) child = g_variant_get_child_value (metadata, i);
       const char *key;
-      g_variant_get_child (child, 0, "&s", &key);
+      g_autoptr(GVariant) value = NULL;
+
+      g_variant_get_child (metadata, i, "{&s@v}", &key, &value);
       if (strcmp (key, "deploy-version") == 0)
         continue;
-      g_variant_builder_add_value (&metadata_builder, child);
+      g_variant_dict_insert_value (&metadata_dict, key, g_steal_pointer (&value));
```

Comment:
This leaks, g_variant_dict_insert_value() only consumes the input variant if it is floating.